plugins/SlicerT/SlicerT.cpp

```diff
@@ -0,0 +1,666 @@
+/*
+ * SlicerT.cpp - simple slicer plugin
+ *
+ * Copyright (c) 2006-2008 Daniel Kauss Serna <daniel.kauss.serna@gmail.com>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "SlicerT.h"
+
+#include <QDomElement>
+#include <fftw3.h>
+#include <math.h>
+
+#include "Engine.h"
+#include "InstrumentTrack.h"
+#include "PathUtil.h"
+#include "Song.h"
+#include "embed.h"
+#include "lmms_constants.h"
+#include "plugin_export.h"
+
+namespace lmms {
+
+extern "C" {
+Plugin::Descriptor PLUGIN_EXPORT slicert_plugin_descriptor = {
+	LMMS_STRINGIFY(PLUGIN_NAME),
+	"SlicerT",
+	QT_TRANSLATE_NOOP("PluginBrowser", "Basic Slicer"),
+	"Daniel Kauss Serna <daniel.kauss.serna@gmail.com>",
+	0x0100,
+	Plugin::Type::Instrument,
+	new PluginPixmapLoader("icon"),
+	nullptr,
+	nullptr,
+};
+} // end extern
+
+PhaseVocoder::PhaseVocoder()
+	: m_FFTInput(s_windowSize, 0)
+	, m_IFFTReconstruction(s_windowSize, 0)
+	, m_allMagnitudes(s_windowSize, 0)
+	, m_allFrequencies(s_windowSize, 0)
+	, m_processedFreq(s_windowSize, 0)
+	, m_processedMagn(s_windowSize, 0)
+{
+	m_fftPlan = fftwf_plan_dft_r2c_1d(s_windowSize, m_FFTInput.data(), m_FFTSpectrum, FFTW_MEASURE);
+	m_ifftPlan = fftwf_plan_dft_c2r_1d(s_windowSize, m_FFTSpectrum, m_IFFTReconstruction.data(), FFTW_MEASURE);
+}
+
+PhaseVocoder::~PhaseVocoder()
+{
+	fftwf_destroy_plan(m_fftPlan);
+	fftwf_destroy_plan(m_ifftPlan);
+}
+
+void PhaseVocoder::loadData(std::vector<float> originalData, int sampleRate, float newRatio)
+{
+	m_dataLock.lock();
+
+	m_originalBuffer = originalData;
+	m_originalSampleRate = sampleRate;
+	m_scaleRatio = -1; // force update, kinda hacky
+
+	m_dataLock.unlock(); // stupid, but QRecursiveMutex is too expensive to have in updateParas and getFrames
```

Comment:
The lock is for the audio thread. It would be bad if the original buffer changed while the audio is being processed. If you remove it, changing the sample during playback gives bad data and corrupts some buffers.

Reply:
> The lock is for the audio thread. It would be bad if the original buffer changed while the audio is being processed. If you remove it, changing the sample during playback gives bad data and corrupts some buffers.

For synchronization with the audio thread, we use ``requestChangesInModel`` and ``doneChangesInModel`` instead. However, please don't actually use those functions, using ``requestChangesGuard`` does the same thing and is wrapped using RAII, which is preferred.

Reply:
Sorry, but how exactly would this work? I don't really understand the internals well enough for this, and I couldn't find any other code that uses `requestChangesGuard`.

Reply:
> Sorry, but how exactly would this work? I don't really understand the internals well enough for this, and I couldn't find any other code that uses requestChangesGuard.

For the RAII mechanism itself, When you create an object from ``requestChangesGuard`` and store it in a variable, the code will call ``requestChangesInModel`` for you when it is created. When the object is destroyed, ``doneChangesInModel`` is called. When we must pair one "starting" action with and "ending" action, like when taking a lock for example, we use RAII. For that lock example I just mentioned, ``std::lock_guard`` would be used, but since the audio thread already has a mechanism for synchronizing changes, we can ``requestChangesGuard`` instead.

For the actual mechanism, ``requestChangesInModel`` will cause the current thread to wait until the audio thread calls ``runChangesInModel``. The audio thread should then allow the changes to be processed sequentially. The thread that is running the change should then call ``doneChangesInModel`` to signify that the change is complete, causing the next one to be processed until there are no more changes left.

Reply:
I tried adding `requestChangesGuard` at the start of `PhaseVocoder::getFrames` like this:
`AudioEngine::RequestChangesGuard audioLock = Engine::audioEngine()->requestChangesGuard();`. 
This is how I understood it should be done. On creation the audio thread is locked, and on function exit it is unlocked.
However, this locks the program almost instantly, which I can only assume is because the lock gets called twice. How would the correct implementation be? Sorry for bothering you with this.

Reply:
> This is how I understood it should be done. On creation the audio thread is locked, and on function exit it is unlocked.

Incorrect. the thread that had called ``requestChangeInModel``/``doneChangeInModel`` (done through ``requestChangesGuard``) will wait, and then the audio thread will control when that thread will wake up.

> However, this locks the program almost instantly, which I can only assume is because the lock gets called twice. How would the correct implementation be? Sorry for bothering you with this.

I assume you had removed ``m_dataLock`` already? If not, my only guess is that ``m_dataLock`` is acquired by the UI thread and then you had called ``requestChangesGuard``, so the change can't run if the audio thread is also trying to acquire ``m_dataLock``. Also, you should only be using ``requestChangesGuard`` in places where the data is actually changing, so not in functions that will be ran by the audio thread.

If ``getFrames`` is called by the audio thread, then the synchronization mechanism shouldn't be used there.

Reply:
> On creation the audio thread is locked, and on function exit it is unlocked.

Incorrect. The thread that had called ``requestChangeInModel``/``doneChangeInModel`` (done through ``requestChangesGuard``) will wait, and then the audio thread will let that thread run the change at the appropriate time.

> However, this locks the program almost instantly, which I can only assume is because the lock gets called twice. How would the correct implementation be? Sorry for bothering you with this.

I assume you had removed ``m_dataLock`` already? If not, my only guess is that ``m_dataLock`` is acquired by the UI thread and then you had called ``requestChangesGuard``, so the change can't run if the audio thread is also trying to acquire ``m_dataLock``. Also, you should only be using ``requestChangesGuard`` in places where the data is actually changing, so not in functions that will be ran by the audio thread.

If ``getFrames`` is called by the audio thread, then the synchronization mechanism shouldn't be used there.

PS: Sorry if you get notified about this message twice, I accidentally started a review and at the same time, wrote this comment on a different tab.

Reply:
Ok, I understand how it works, and as I see it I can't add this guard. Because `getFrames` has to be thread safe in the audio thread and the UI thread. 

This is because `getFrames` dynamically calculates the time shifted sample, and for that it uses the buffers that are members of the class. If you were to call `getFrames` from two notes simultaneously for example, the result would be a complete mess, since the buffers would be written to from 2 audio threads. And of course they also have to me locked for the UI thread.

This is another drawback of the problems I talked about in this [comment](https://github.com/LMMS/lmms/pull/6857#issuecomment-1769362384).

> PS: Sorry if you get notified about this message twice, I accidentally started a review and at the same time, wrote this comment on a different tab.

I actually didn't get notified at all for some reason, and only just saw it while checking if there was something left to do. Probably a problem with my notification settings or something.